test/functional.jl

```diff
@@ -239,3 +239,33 @@ end
 @test_throws DimensionMismatch Base.IteratorND(1:2, (2,3))
 
 @test collect(Base.Generator(+, [1,2], [10,20])) == [11,22]
+
+# partition(c, n)
+
+let v = collect(Base.partition([1,2,3,4,5], 1))
+    @test all(i->v[i][1] == i, v)
+end
+
+let v = collect(Base.partition([1,2,3,4,5], 2))
+    @test v[1] == [1,2]
+    @test v[2] == [3,4]
+    @test v[3] == [5]
+end
+
+let v = collect(Base.partition(enumerate([1,2,3,4,5]), 3))
+    @test v[1] == [(1,1),(2,2),(3,3)]
+    @test v[2] == [(4,4),(5,5)]
+end
+
+for n in [5,6]
+    @test collect(Base.partition([1,2,3,4,5], n))[1] == [1,2,3,4,5]
+    @test collect(Base.partition(enumerate([1,2,3,4,5]), n))[1] ==
+          [(1,1),(2,2),(3,3),(4,4),(5,5)]
+end
+
+
+@test_throws AssertionError Base.partition([1,2,3,4,5], 0)
+@test_throws AssertionError Base.partition([1,2,3,4,5], -1)
+
+@test join(map(x->string(x...), Base.partition("Hello World!", 5)), "|") ==
```

Comment:
This seems to work fine now that it is not using character indexing (it is using character iteration now).
See this example where `partition` does the expected thing with the UTF-8 emoji, but string indexing causes an error:

`foo.jl`:
```julia
s = "Monkey 🙈🙊🙊"
@show s
v = collect(Base.partition(s, 4))
@show v
v = collect(Base.partition(s, 3))
@show v
v = collect(Base.partition(s, 2))
@show v
v = collect(Base.partition(s, 1))
@show v

v = collect(Base.partition(s, 8))
@show v

@show s[8]
@show s[12]
@show try s[9] catch e e end
```

```bash
$ ./julia foo.jl
s = "Monkey 🙈🙊🙊"
v = [['M','o','n','k'],['e','y',' ','🙈'],['🙊','🙊']]
v = [['M','o','n'],['k','e','y'],[' ','🙈','🙊'],['🙊']]
v = [['M','o'],['n','k'],['e','y'],[' ','🙈'],['🙊','🙊']]
v = [['M'],['o'],['n'],['k'],['e'],['y'],[' '],['🙈'],['🙊'],['🙊']]
v = [['M','o','n','k','e','y',' ','🙈'],['🙊','🙊']]
s[8] = '🙈'
s[12] = '🙊'
try s[9] catch e e end = UnicodeError: invalid character index
```